File: Leetcode_POTD_2025/8_August/3479_Fruits_Into_Baskets_III.java
import java.util.Arrays;

class Solution {
    public int numOfUnplacedFruits(int[] fruits, int[] baskets) {
        int numBaskets = baskets.length;
        int treeSize = 1;
        while (treeSize < numBaskets) {
            treeSize *= 2;
        }

        int[] segmentTree = new int[2 * treeSize];
        Arrays.fill(segmentTree, -1);
        for (int i = 0; i < numBaskets; ++i) {
            segmentTree[treeSize + i] = baskets[i];
        }

        for (int i = treeSize - 1; i > 0; --i) {
            segmentTree[i] = Math.max(segmentTree[2 * i], segmentTree[2 * i + 1]);
        }

        int unplacedCount = 0;
        for (int fruitQty : fruits) {
            if (segmentTree[1] < fruitQty) {
                unplacedCount++;
                continue;
            }

            int currentNodeIndex = 1;
            while (currentNodeIndex < treeSize) {
                if (segmentTree[2 * currentNodeIndex] >= fruitQty) {
                    currentNodeIndex = 2 * currentNodeIndex;
                } else {
                    currentNodeIndex = 2 * currentNodeIndex + 1;
                }
            }

            segmentTree[currentNodeIndex] = -1;

            currentNodeIndex /= 2;
            while (currentNodeIndex > 0) {
                segmentTree[currentNodeIndex] = Math.max(segmentTree[2 * currentNodeIndex], segmentTree[2 * currentNodeIndex + 1]);
                currentNodeIndex /= 2;
            }
        }
        return unplacedCount;
    }
}

File: Leetcode_POTD_2025/8_August/3479_Fruits_Into_Baskets_III.py
import math

class Solution(object):
    def numOfUnplacedFruits(self, fruits, baskets):
        """
        :type fruits: List[int]
        :type baskets: List[int]
        :rtype: int
        """
        num_baskets = len(baskets)
        tree_size = 1
        while tree_size < num_baskets:
            tree_size *= 2

        segment_tree = [-1] * (2 * tree_size)
        
        for i in range(num_baskets):
            segment_tree[tree_size + i] = baskets[i]

        for i in range(tree_size - 1, 0, -1):
            segment_tree[i] = max(segment_tree[2 * i], segment_tree[2 * i + 1])

        unplaced_count = 0
        for fruit_qty in fruits:
            if segment_tree[1] < fruit_qty:
                unplaced_count += 1
                continue

            current_node_index = 1
            while current_node_index < tree_size:
                if segment_tree[2 * current_node_index] >= fruit_qty:
                    current_node_index = 2 * current_node_index
                else:
                    current_node_index = 2 * current_node_index + 1
            
            segment_tree[current_node_index] = -1

            current_node_index //= 2
            while current_node_index > 0:
                segment_tree[current_node_index] = max(segment_tree[2 * current_node_index], segment_tree[2 * current_node_index + 1])
                current_node_index //= 2
                
        return unplaced_count
